### DISCUSSION TITLE
Circles do not render in any browser, everything else does.

### DISCUSSION THREAD

**Discussion by claudioap** [maplibre/maplibre-gl-js#4402] [Q&A]
See this example? https://maplibre.org/maplibre-gl-js/docs/examples/drag-a-point/
Good, however I do not!

Maplibre just stopped drawing circles in my laptop. Firefox, Brave, Edge, Chromium; happens in all of them.  (OS is Linux but it also happens inside of Flatpaks, it is not limited to native code).

I can draw anything else on maps (lines, text, ...) but circles. Note that handlers work, so I can click and interact with the invisible circles, I simply do not see them. No errors, atypical-warnings or anything that points to problems.

I thought this was something with my code but I can see it happening in the maplibre examples. There are however some pages that do use maplibre where I do not see that happening, for example, here: https://beta.carrismetropolitana.pt/stops/all

One of my map usages still works fine; in there I gave different colors to points but I can't reproduce it. Tried to just recolor a point or interpolate the color based on things like zoom and no luck. Also tried to defer data insertion and layer creation to way after the map is loaded. No luck either.

Any ideas?

**Comment by HarelM** [comment]
Have you tried older versions of maplibre? It might be an issue with the webgl driver for webgl 2, maybe, IDK...

**Reply by claudioap** [reply]
It happens as of 4.0 and 4.6.

The computer that can reproduce the issue has an AMD 5825U
Exact same setup with a 3700X + GTX 1070  works fine.
The 5825U already rendered these things fine in the past so, if the regression has anything to do with Maplibre, it is not only about Maplibre. Something else had to happen.

Perhaps, anyone with the same RDNA graphics version running the same Flatpaks can reproduce the issue.
Perhaps, it is contained to my computer, albeit it is weird that is literally only affects circles drawn in a map library.

The way the issues is posed, I'm considering that maybe there's some corrupted lib or troubled graphics driver. All of these browsers have close to nothing in common. Some of them are using my system native libs, some of them are using Flatpak's base runtimes, some of them are Gecko-based, some of them are Blink-based.

I'll try to downgrade the GPU driver.

**Reply by claudioap** [reply]
This example only makes things weirder: https://maplibre.org/maplibre-gl-js/docs/examples/cluster/ I see the clusters, just not the individual circles.
![image](https://github.com/user-attachments/assets/a6eb8909-5226-4703-9ba6-da23d66a6561)
![image](https://github.com/user-attachments/assets/8078cacb-01da-482b-9394-b31ed6210b96)
![image](https://github.com/user-attachments/assets/565fa0b2-b535-4e27-ac8d-29736a8ff81b)
![image](https://github.com/user-attachments/assets/215f4313-c7a9-4d11-8d68-7cb319cd9feb)


I already downgraded all of my system to the packages I had as of Jun 1st (and at the time I'm pretty sure this worked) and still no luck.

What an odd and weird bug.

**Reply by claudioap** [reply]
Some additional information: I kept blind developing and  I eventually saw points. One of my layers partially drew most of its points. Not all, just some.

This happened as I added another entirely independent source with its layers (also drawing points, 100% invisible). It was a rather big source if it matters. ~60k points which get "drawn" before (so, lower) than the layer that is getting a partial draw.

**Reply by filip-769** [reply]
I can reproduce this exact same issue. It was working before and now it doesn't.
I tried firefox, ungoogled chromium and gnome web. It works fine on my android phone.

# System Details Report
---

## Report details
- **Date generated:**                              2024-08-15 13:07:36

## Hardware Information:
- **Hardware Model:**                              ASUSTeK COMPUTER INC. Vivobook_ASUSLaptop M1605YA_M1605YA
- **Memory:**                                      8,0 GiB
- **Processor:**                                   AMD Ryzen™ 5 7530U with Radeon™ Graphics × 12
- **Graphics:**                                    AMD Radeon™ Graphics
- **Disk Capacity:**                               (null)

## Software Information:
- **Firmware Version:**                            M1605YA.302
- **OS Name:**                                     Arch Linux
- **OS Build:**                                    rolling
- **OS Type:**                                     64-bit
- **GNOME Version:**                               46
- **Windowing System:**                            Wayland
- **Kernel Version:**                              Linux 6.10.4-arch2-1

**Reply by cutephoton** [reply]
This fixes it: https://github.com/BananaSlugLabs/maplibre-gl-js/commit/0764607fef7c4bb5867c4d5a4c8e574048d767b4

Unfortunately I have no idea why this works.

**Reply by HarelM** [reply]
@ibesora @kubapelc can you guys guess why the above code change solves this issue?

**Reply by ibesora** [reply]
I can see why this would fix the issue, but it's just masking the underlying issue. 
I'd guess there's something going on with the vertex buffer.
@claudioap @filip-769 can I ask you to go to [this](https://jsfiddle.net/ibesora/0Ld9xeu8/8/) fiddle and see if you get a grey box similar to the one on the center of the image below?

<img width="734" alt="Screenshot 2024-09-05 at 17 05 09" src="https://github.com/user-attachments/assets/01af3242-40b6-4810-8739-94a060cb50f7">

**Reply by cutephoton** [reply]
If somebody has some suggestions for where I might look or something to check/try, I can explore this issue further.

`length(v_data.xyz)` also worked which, as above, is probably doing nothing more than masking the root cause.

I could not reproduce this issue on intel i-gpu. I was able to reproduce on AMD GPU (as others have). A comment above suggests nvidia also work fine, but I could not test.

I explored mesa's repo. This appeared first in Mesa 24.1.0 based on the history in the bug report. This was a big delta so hard to identify anything specific. I did locate a recent, brief, but [interesting issue](https://gitlab.freedesktop.org/mesa/mesa/-/issues/11513). It's possible that some optimizations, under some circumstances, can yield different results if using a uniform due to slight rounding errors caused by a different order of operations. Since the issue appears only when a uniform is used for color, I wonder if something similar is happening that is specific to AMD GPUs. However, the discussion was about atomic optimizations which isn't related. (I thought I would clarify that I think it's a very low probability since this is much more than a slight rounding error.)

**Reply by cutephoton** [reply]
So after meandering down some weird paths and experiencing some weird behaviors in the fragment shader that didn't make much sense, I decided to explore the vertex shader. The antialiasingblur value makes sense now. Z is calculated by `(1.0 / u_device_pixel_ratio / (radius + stroke_width))` which, per comment, is approximately 1px. Dumping `circleUniformValues` from js shows that this value, on my machine, is 1.25 (or about 0.008 for a radius of 100).

`lowp float antialiasblur = (1.0 / u_device_pixel_ratio / (radius + stroke_width));` -> Does not work.
`lowp float antialiasblur = 0.008;` -> Works
`lowp float antialiasblur = (1.0 / 1.25 / (radius + stroke_width));` -> Works
`lowp float antialiasblur = (1.0 / (radius + stroke_width));` -> Works (obviously)
`lowp float antialiasblur = (1.0 / u_device_pixel_ratio / radius);` -> Does not work.
`lowp float antialiasblur = (1.0 / u_device_pixel_ratio);` -> Works
`lowp float antialiasblur = 1.0 / u_device_pixel_ratio / 10.0;` -> Works (Radius = 10)

This is obviously does not compute the correct value, but it renders the circle:

`lowp float antialiasblur = 1.0 / (u_device_pixel_ratio + radius);`

When the expression is in the form of `1 / var / var`, the circle fails to draw? Any other combination works `literal / literal / var` or `literal / var / literal`.

Somebody is gonna have to make this make sense. :) 

<details>
  <summary>A lengthy misadventure with fragments that ends in confusion.</summary>
  Shaders aren't my specialty but strap in and follow along as I take you on an adventure...

```
void main() {
    // ...
    fragColor = vec4(
            length(v_data.xy),
            0,
            0,
            1.0);
}
```

![image](https://github.com/user-attachments/assets/f0668888-4c62-4469-895b-b44acd1818f2)

So lets add the smoothstep result to the B channel:

```
fragColor = vec4(
            length(v_data.xy),
            0.0,//v_data.z,
            smoothstep(0.0, -v_data.z, length(v_data.xy) - 1.0),
            1.0);
```

![image](https://github.com/user-attachments/assets/87450115-c309-4a5b-b91f-4de47e2e9704)

Hmmm! Solid color. This renders correctly on Intel IGP. Lets try something completely random that shouldn't fix this.

```
fragColor = vec4(
            length(v_data.xy),
            v_data.z,
            smoothstep(0.0, -v_data.z, length(v_data.xy) - 1.0),
            1.0);
```

![image](https://github.com/user-attachments/assets/188cce4a-8daf-4167-af01-2cd4a28a6256)

Weird. Setting the G component to `v_data.z` magically causes things to start working as expected. (Z in this case is 0x02 or about 0.0078.) Lets try something different:

```
fragColor = vec4(
            length(v_data.xy),
            0.0 >= -v_data.z ? 0.25 : 1.0, // <<<<< Is edge0 >= edge1?
            smoothstep(0.0, -v_data.z, length(v_data.xy) - 1.0),
            1.0);
```

![image](https://github.com/user-attachments/assets/628fa9ef-d1e3-4caa-a0b6-3fc18fd5384c)

Oh! The green channel is 0x40 (0.25). edge0 >= edge1 is true. And that brings me back to the GLSL specification:

![image](https://github.com/user-attachments/assets/e3574e7c-8e68-40f7-a3d8-63d3c85d2adc)

Bingo! Let's fix it!

```
    fragColor = vec4(
        length(v_data.xy),
        0.0 >= -v_data.z ? 0.25 : 1.0, // <<<<< Is edge0 >= edge1?
        smoothstep(-v_data.z, 0.0,  1.0 - length(v_data.xy)),
        1.0);
```

![image](https://github.com/user-attachments/assets/eb84f6de-f8ba-4415-ad1c-7ad0947c4dfd)

Ok! Time to circle back to making a proper circle:

```
fragColor = color * smoothstep(-v_data.z, 0.0,  1.0 - length(v_data.xy));
```

![image](https://github.com/user-attachments/assets/55fd4349-b631-4043-ac5c-234130a275d6)

Look at that! So now that this is sorted out, start moving back to the original shader. (It doesn't go as planned.)

```
float opacity_t = smoothstep(-max(blur, v_data.z), 0.0,  1.0 - length(v_data.xy));
    fragColor = v_visibility * opacity_t * color;
```

![image](https://github.com/user-attachments/assets/8ec774f2-e203-4719-ae81-35e656cb7b53)

Huh? I'm lost again. So reverting back to some good old guess 'n check code...

```
    float opacity_t = -max(blur, v_data.z) >= 0.0 ? 0.0 : smoothstep(min(-blur, -v_data.z), 0.0,  1.0 - length(v_data.xy));
    fragColor = v_visibility * opacity_t * color;
```

![image](https://github.com/user-attachments/assets/e88d7922-9a43-4136-8ccd-d44bd5e59607)

While this...
```
    float opacity_t = -max(blur, v_data.z) >= 0.0 ? 0.0 : smoothstep(-max(blur, v_data.z), 0.0,  1.0 - length(v_data.xy));
    fragColor = v_visibility * opacity_t * color;
```
Doesn't work. Setting aside corner cases (NaNs and Inf), `-max(blur, v_data.z)` should be equiv for `min(-blur, -v_data.z)`. However, I think this is a red herring. Earlier, I added a value (i.e. an extra instruction) to the output color vector which caused the `smoothstep` function to work as expected even though the output was, per the spec, undefined. I need to investigate further.

(Edit)
For some reason, `v_data.z` always seems to un-break the shader when added somewhere in the color vector. Using some other value like, for example, `blur` doesn't have the same effect.
```
float opacity_t = smoothstep(-max(blur, v_data.z), 0.0,  1.0 - length(v_data.xy));
fragColor = v_visibility * opacity_t * vec4(v_data.z + color.x, color.y, color.z, color.w);
```
It does not work for x or y. As well, `v_visibility * opacity_t * color * (1-v_data.z)` did not work. Maybe smoothstep issue was a red herring too. Something funky is going on deep down inside.

Random observation: On AMD GPU, `v_data.z` ~= 0.0078. On Intel IGP, `v_data.z` is ~= 0.01. (0x2 vs 0x3)
(/Edit)

I'll continue exploring in future comment. Please let me know if I've got something wrong along the way. It's very plausible that I have!
</details>

**Reply by cutephoton** [reply]
Thoughts on this as a fix? It dawned on me that we don't need to calculate this parameter in the fragment shader. (I don't love that I wasn't able to sort out a root cause. By all appearances this looks like a bug in mesa/amd?)

https://github.com/BananaSlugLabs/maplibre-gl-js/commit/4bcfea25b5577ea5dbc1be9f5aab6c1676d4e331

![image](https://github.com/user-attachments/assets/a14a9666-101e-4a94-b38d-b8ae2bba1130)

**Reply by HarelM** [reply]
It might be some optimization done incorrectly in these specific cases...?

**Reply by cutephoton** [reply]
Seems to be the most plausible explanation at this time. (I am surprised to see it happen on something this uncomplicated.) Should I do a PR? I am loath to push workarounds, though the change avoids recalculating the same value repeatedly so maybe it has some technical merit.

**Reply by HarelM** [reply]
Please open a PR. This is not reproducible in our current CI, right?

**Reply by cutephoton** [reply]
Can be reproduced with `npm run test-render -- circle-radius --open-browser` on impacted machines, but it would not reproduce headless.

![Screenshot from 2024-09-10 14-36-01](https://github.com/user-attachments/assets/d4579c3d-a153-4ee5-94de-625bf7f6c88e)

**Reply by HarelM** [reply]
I don't think we have impacted machines as part of the CI pool though...
